### PR TITLE
Remove gpio dispatch macro-defining proc macro

### PR DESCRIPTION
**Pull request by bugadani**
I wasn't able to decipher how the proc macro worked so I dropped it 🤪🤯

Also includes some simplifications.

### CODE REVIEW

**Comment by bugadani** [comment]
Oh hmm I should have tested this with something other than S3 :/

**Comment by bugadani** [comment]
I'm going to assume HIL failures are unrelated.

File: esp-hal/src/gpio/mod.rs [hunk 1]
```diff
@@ -645,19 +645,34 @@ pub fn connect_high_to_peripheral(signal: InputSignal) {
...
+pub struct False {}
+
+impl BooleanType for False {
+    const VALUE: bool = false;
+}
 
 #[doc(hidden)]
-pub trait IsTouchPin: PinType {}
+pub trait PinType {
+    type IsOutput: BooleanType;
```
Comment (bjoernQ) [inline esp-hal/src/gpio/mod.rs:668]: I like this idea

Comment (bugadani) [reply esp-hal/src/gpio/mod.rs:668]: I'm a bit worried if this is just more complicated than the marker traits, for no good reason.

**Review by bjoernQ — APPROVED** [review]
Yeah, the enum_dispatch thing was confusing - the replacement and `if_output_pin!` is much easier to understand

**Review by JurajSadel — APPROVED** [review]
Thanks!

**Review by jessebraham — APPROVED** [review]
Thanks!

**Pull request merged**